### DISCUSSION TITLE
airgap image installation containerd

### DISCUSSION THREAD

**Discussion by cwrx777** [k3s-io/k3s#9497] [Q&A] [answered]
Hi,

my environment is rhel 8.9 and I want to use containerd. 
since containerd is bundled in k3s, in which container engine the airgap image is [loaded](https://docs.k3s.io/installation/airgap#load-images) ?

**Comment by brandond** [comment]
I'm confused by your question. Airgap image preloading is only supported with the embedded containerd. If you leave the default configuration in place - don't enable `--docker` to use cri-dockerd, or `--container-runtime-endpoint` to use your own runtime socket - then you're using the embedded containerd. Is there some other specific detail you're looking for?

**Reply by cwrx777** [reply]
@brandond 
Thank you for your reply.

I'm trying to understand the installation steps of k3s in airgap environment in RHEL 8.x. In my understanding of the steps documented [here](https://docs.k3s.io/installation/airgap), as [prerequisites](https://docs.k3s.io/installation/airgap#prerequisites), we need to prepopulate the images that k3s needs to install using either of the two methods. 
1. For the [private registry method](https://docs.k3s.io/installation/airgap#private-registry-methodl), 
    - if the private registry is also in the master node, and it does not have docker, is it possible to use the containerd that is bundled with k3s to run the private registry?
    - or firstly I need to install containerd in the master node to run the private registry and then install k3s? 
2. For [manually deploy method](https://docs.k3s.io/installation/airgap#manually-deploy-images-method), 
    - does this mean that we don't need to have a running container engine in the master node?
    - I intend to use different [`data-dir`](https://docs.k3s.io/cli/server#data) when running the server, and I notice in this installation method the [airgap image is saved in the default `data-dir](https://docs.k3s.io/installation/airgap#prepare-the-images-directory-and-airgap-image-tarball)`. is it possible to specify different `data-dir` for this installation method?

> "Airgap image preloading is only supported with the embedded containerd"

Can you also explain what you meant the above ? any links or example to run the install script will be good.

**Reply by brandond** [reply]
No, the registry can't be run in k3s's embedded containerd... because containerd isn't up until k3s is up. It's part of k3s and you can't use to host things before k3s itself is running. 

If you don't have an existing registry to push images to then:

1. Put your image tarballs in `$DATADIR/agent/images/`
1. Run the install script with `--data-dir=$DATADIR`

You might also take a look at https://docs.k3s.io/installation/registry-mirror - if you enable and configure this, then you only need the tarballs on one node; the other ones can copy the images from the first.

**Reply by cwrx777** [reply]
Hi @brandond,

> No, the registry can't be run in k3s's embedded containerd... because containerd isn't up until k3s is up. It's part of k3s and you can't use to host things before k3s itself is running.

is it possible to run the private registry like Harbor in the k3s cluster and configure Harbor only to run in the master node. This private registry is meant for the worker to pull the images from.

**Reply by brandond** [reply]
Possibly, but bootstrapping harbor into an airgap environment will be a fair bit of work. You're probably better off using the embedded registry mirror that I linked earlier, have you looked at that yet?

**Reply by cwrx777** [reply]
@brandond 
My understanding on the setup is as follows, let me know if I'm missing something/incorrect. 
1. each node `registries.yaml` needs to have the `mirrors` section.
2. Load the image tarball by using the `ctr -n k8s.io image import ` command. Then all the images will be automatically mirrored/replicated in all nodes.

**Reply by brandond** [reply]
if you put things in the images dir before starting k3s, you don't need to *also* manually import them. The archives in the images dir are only imported on startup, but you can add **more** things later, while k3s is running, by manually importing them.

**Reply by cwrx777** [reply]
@brandond 

I've added some entries in `registries.yaml` in both master and worker nodes, just as follows, without any other configurations:
```yaml
mirrors:
  quay.io:
  docker.io:
  registry.k8s.io:
  gcr.io:
```

In the worker node, the agent service was not able to restart and its stuck with message:
```
Waiting to retrieve agent configuration; server is not ready: failed to get CA certs: Get \"https://127.0.0.1:6444/cacerts
failed to retrieve agent configuration: failed to get CA certs: Get \"https://127.0.0.1:6444/cacerts\
```
In the embedded registry docs, there is [authentication](https://docs.k3s.io/installation/registry-mirror#authentication) section, which I am sure how to set up. Is it related to the [private registry config](https://docs.k3s.io/installation/private-registry#configs) ?

the worker node can reach the server's API url
```bash
curl -vs https://20.0.20.196:6443
*   Trying 20.0.20.196...
* TCP_NODELAY set
* Connected to 20.0.20.196 (20.0.20.196) port 6443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
 ```

**Reply by brandond** [reply]
Is the server running? Are you sure you've specified the server address correctly on the agent? Have you disabled firewalls between the two nodes, or opened the correct ports? The error you're getting has nothing to do with mirror configuration.

Note that you're not actually showing a response to your curl test...

**Reply by cwrx777** [reply]
@brandond ,
What does this [authentication ](https://docs.k3s.io/installation/registry-mirror#authentication)section mean? Is it required? and how to configure if it's required.

**Reply by brandond** [reply]
You don't need to do anything about that, no. If a node is part of the cluster then authentication to the registry API on other cluster members is handled for you.

Is this at all related to the previous question you asked? Did you get that resolved?

**Reply by cwrx777** [reply]
Hi @brandond , 
You were right.. it's security group issue. now the agent can curl the server's cacert URL. 
However, even after I restarted server and agent k3s services, I didn't see that the images were in the agent node (`sudo ctr image ls -q`) .
I can see there is connection from agent to server at port 5001. 
when do we expect to see that the images were mirrored to other nodes? or is it only pulled when it's needed?

**Reply by brandond** [reply, marked as answer]
https://docs.k3s.io/installation/registry-mirror#pushing-images
> Note that the k8s.io namespace must be specified when managing images via ctr in order for them to be visible to the kubelet.

have you tried doing `ctr -n k8s.io image ls` ?

Images are not actively pushed to other nodes. they are pulled from nodes that have the image, by nodes that need the image, when a pull is initiated by a user or the kubelet.

**Reply by cwrx777** [reply]
Hi @brandond,

> Images are not actively pushed to other nodes. they are pulled from nodes that have the image, by nodes that need the image, when a pull is initiated by a user or the kubelet.

Thanks for the explanation. This is what I observed.
